homeassistant/components/lcn/helpers.py

```diff
@@ -1,18 +1,191 @@
+        device_config[CONF_IS_GROUP],
+    )
+
+
+def get_device_connection(hass, unique_device_id, config_entry):
+    """Return a lcn device_connection."""
+    device_config = get_device_config(unique_device_id, config_entry)
+    if device_config:
+        host_connection = hass.data[DOMAIN][config_entry.entry_id][CONNECTION]
+        addr = pypck.lcn_addr.LcnAddr(*get_device_address(device_config))
+        device_connection = host_connection.get_address_conn(addr)
+        return device_connection
+    return None
+
+
+def get_resource(domain_name, domain_data):
+    """Return the reosurce for the specified domain_data."""
+    if domain_name in ["switch", "light"]:
+        return f'{domain_data["output"]}'
+    if domain_name in ["binary_sensor", "sensor"]:
+        return f'{domain_data["source"]}'
+    if domain_name == "cover":
+        return f'{domain_data["motor"]}'
+    if domain_name == "climate":
+        return f'{domain_data["source"]}.{domain_data["setpoint"]}'
+    if domain_name == "scene":
+        return f'{domain_data["register"]}.{domain_data["scene"]}'
+    raise ValueError("Unknown domain.")
+
+
+def generate_unique_id(
+    address=None,
+    domain_config=None,
+):
+    """Generate a unique_id from the given parameters."""
+    unique_id = ""
+    if address:
+        is_group = "g" if address[2] else "m"
+        unique_id += f"{is_group}{address[0]:03d}{address[1]:03d}"
+        if domain_config:
+            resource = get_resource(*domain_config)
+            unique_id += f"-{resource}".lower()
+    return unique_id
+
+
+def import_lcn_config(lcn_config):
+    """Convert lcn settings from configuration.yaml to config_entries data."""
+    data = {}
+    for connection in lcn_config[CONF_CONNECTIONS]:
+        host = {
+            CONF_HOST: connection[CONF_NAME],
+            CONF_IP_ADDRESS: connection[CONF_HOST],
+            CONF_PORT: connection[CONF_PORT],
+            CONF_USERNAME: connection[CONF_USERNAME],
+            CONF_PASSWORD: connection[CONF_PASSWORD],
+            CONF_SK_NUM_TRIES: connection[CONF_SK_NUM_TRIES],
+            CONF_DIM_MODE: connection[CONF_DIM_MODE],
+            CONF_DEVICES: [],
+            CONF_ENTITIES: [],
+        }
+        data[connection[CONF_NAME]] = host
+
+    for domain, domain_config in lcn_config.items():
+        if domain == CONF_CONNECTIONS:
+            continue
+        domain_name = DOMAIN_LOOKUP[domain]
+        # loop over entities in configuration.yaml
+        for domain_data in domain_config:
+            # remove name and address from domain_data
+            entity_name = domain_data.pop(CONF_NAME)
+            address, host_name = domain_data.pop(CONF_ADDRESS)
+
+            if host_name is None:
+                host_name = DEFAULT_NAME
+
+            # check if we have a new device config
+            unique_device_id = generate_unique_id(address)
+            for device_config in data[host_name][CONF_DEVICES]:
```

Comment:
Exactly that point gave me a lot of headaches when defining the data structure for the config_entry. :smirk:
Hard to explain, but basically it's a trade-off.

When setting up the platforms/entities (and when sending data via websocket api to the frontend later on), I need all the data for an entitiy/device stored in the list entries (which are basically dicts). 
Example:
```python
    "entities": [
      {
        "unique_device_id": "m000007",
        "name": "Light_Output1",
        "resource": "output1",
        "domain": "light",
        "domain_data": {
          "output": "OUTPUT1",
          "dimmable": true,
          "transition": 5000.0
        }, ...
```

Using some of the entry values as dict keys, would mean that we always have to pass around key/value pairs instead of a self-consistent dict for a device/entity to have all the data we need at hand. It turned out that this will be complicated to realize in a readable way.
Example:
```python
    "entities": {
        ("m000007", "output1", "light"):
          {
            "name": "Light_Output1",
            "domain_data": {
              "output": "OUTPUT1",
              "dimmable": true,
              "transition": 5000.0
            }
        },
    }
```

One solution would be to **not** remove the values from the entry we use as dict keys. But that would mean that we have redundant data in the config_entry. I don't like that.
Example:
```python
    "entities": {
        ("m000007", "output1", "light"):
          {
            "unique_device_id": "m000007",
            "name": "Light_Output1",
            "resource": "output1",
            "domain": "light",
            "domain_data": {
              "output": "OUTPUT1",
              "dimmable": true,
              "transition": 5000.0
            }
        },
   }

```



So, that's basically the reason why I chose lists for the devices and entities along with two convenient methods to get the right configuration data out. I think it pays off later.

Reply:
Can you point me to where the lists are needed? I can't quite find it.

Reply:
When setting up an entity I need to pass the entity_config dictionary (example binary_sensor):
https://github.com/home-assistant/core/blob/0ced19eae4ef773539631e3d38f7c94015d5a980/homeassistant/components/lcn/binary_sensor.py#L43

It's then used to setup the connection to the hardware device:
https://github.com/home-assistant/core/blob/0ced19eae4ef773539631e3d38f7c94015d5a980/homeassistant/components/lcn/binary_sensor.py#L21-L26

And finally used in the entity instance. For example to generate the unique_id:
https://github.com/home-assistant/core/blob/0ced19eae4ef773539631e3d38f7c94015d5a980/homeassistant/components/lcn/__init__.py#L183-L189

I think it makes more sense to have a list of dicts which store all entity attributes, rather than storing some of the attributes them in a key and some of them in a value of a superior dict.
The list as it is will be needed as soon as the device/entity configurations are send to the frontend via websocket api.

Reply:
OK -- I think where its hard to follow is in the getter methods above, so i may re-review those or make other suggestions to simplify.

Reply:
Maybe my suggestion is really just to have an additional index from address to config entry id  either directly in hass.data or from a class stored in hass.data for managing the common lookups to avoid the deep call stack with many parameters.   Something to consider for future PRs.

Reply:
A class which exposes config_entry.data with a nice interface to ease the access  might be of help. But it could also make things more complicated because everything has to be kept nicely in sync (entry.data, deviceregistry, entityregistry) and everything needs to be triggered vice versa. I would like to postpone it to future optimizations.

Reply:
Agreed "Something to consider for future PRs".   My impression is we already have to keep entry.data, device registry, entity registry in sync, so not sure it makes it harder, but I agree I would not do it if readability is not improved.  One way to think about it is a separation of responsibilities: Consider what the "API" would look like if the device config came from the device itself.  I realize we're edging into style preferences here, so just something to think about.